z3/src/probe.rs

```diff
@@ -0,0 +1,218 @@
+    pub fn ge(&self, p: &Probe) -> Probe<'ctx> {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            let z3_probe = Z3_probe_ge(self.ctx.z3_ctx, self.z3_probe, p.z3_probe);
+            Z3_probe_inc_ref(self.ctx.z3_ctx, z3_probe);
+            Probe {
+                ctx: self.ctx,
+                z3_probe,
+            }
+        }
+    }
+
+    /// Return a probe that evaluates to "true" when the value returned by `self` is equal to the value returned by `p`.
+    pub fn eq(&self, p: &Probe) -> Probe<'ctx> {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            let z3_probe = Z3_probe_eq(self.ctx.z3_ctx, self.z3_probe, p.z3_probe);
+            Z3_probe_inc_ref(self.ctx.z3_ctx, z3_probe);
+            Probe {
+                ctx: self.ctx,
+                z3_probe,
+            }
+        }
+    }
+
+    /// Return a probe that evaluates to "true" when `self` and `p` evaluates to true.
+    pub fn and(&self, p: &Probe) -> Probe<'ctx> {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            let z3_probe = Z3_probe_and(self.ctx.z3_ctx, self.z3_probe, p.z3_probe);
+            Z3_probe_inc_ref(self.ctx.z3_ctx, z3_probe);
+            Probe {
+                ctx: self.ctx,
+                z3_probe,
+            }
+        }
+    }
+
+    /// Return a probe that evaluates to "true" when `p1` or `p2` evaluates to true.
+    pub fn or(&self, p: &Probe) -> Probe<'ctx> {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            let z3_probe = Z3_probe_or(self.ctx.z3_ctx, self.z3_probe, p.z3_probe);
+            Z3_probe_inc_ref(self.ctx.z3_ctx, z3_probe);
+            Probe {
+                ctx: self.ctx,
+                z3_probe,
+            }
+        }
+    }
+
+    /// Return a probe that evaluates to "true" when `p` does not evaluate to true.
+    pub fn not(&self) -> Probe<'ctx> {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            let z3_probe = Z3_probe_not(self.ctx.z3_ctx, self.z3_probe);
+            Z3_probe_inc_ref(self.ctx.z3_ctx, z3_probe);
+            Probe {
+                ctx: self.ctx,
+                z3_probe,
+            }
+        }
+    }
+
+    /// Return a probe that evaluates to "true" when the value returned by `self` is not equal to the value returned by `p`.
+    pub fn ne(&self, p: &Probe) -> Probe<'ctx> {
+        unsafe {
+            let guard = Z3_MUTEX.lock().unwrap();
+            let z3_probe = Z3_probe_eq(self.ctx.z3_ctx, self.z3_probe, p.z3_probe);
+            Z3_probe_inc_ref(self.ctx.z3_ctx, z3_probe);
+            let z3_probe_not = Z3_probe_not(self.ctx.z3_ctx, z3_probe);
+            Z3_probe_inc_ref(self.ctx.z3_ctx, z3_probe_not);
+            Probe {
+                ctx: self.ctx,
+                z3_probe: z3_probe_not,
+            }
+        }
+    }
+}
```

Comment:
These methods really lend themselves to easy [std::ops](https://doc.rust-lang.org/std/ops/) operator implementation over Probe. Would you be able to add that if you have time?

I'm imagining a scenario like
```rust
let probe_a;
let probe_b;
// Nice ergonomics!
probe_a < probe_b
```

Reply:
yeah I was actually considering that.  I've never used `std::ops` so this is a good time to learn!  I'll try to add that in tomorrow.

Reply:
at first glance though, it looks like something like this would work:

```
impl<'ctx> Not for Probe<'ctx> {
    type Output = Probe<'ctx>;

    fn not(self) -> Self::Output {
        self.not()
    }
}
```

Since you can overload std::ops, but `std::cmp::PartialEq` seems to require a return type of `bool` : https://doc.rust-lang.org/src/core/cmp.rs.html#203

Reply:
Maybe adding `std::ops` support could be another PR?  I think those can be a bit more contentious, while exposing core Z3 functionality seems like a straight win.

Reply:
sounds good to me